Fix GLM-4.7 MoE Detector complex JSON Schema type parsing

Leoyzen opened the pull request:
- Modify get_argument_type function to correctly parse complex JSON Schema structures including anyOf/oneOf/allOf and OpenAI-style array types like ["string", "null"]
- Improve _format_value_complete function to properly handle object/array types as JSON rather than strings
- Add additional parsing strategies in parse_arguments for object/array type handling
- Add comprehensive test cases to verify complex type definitions are correctly parsed as their appropriate data types (e.g., arrays) rather than strings

This fixes the issue where array values like [] were being parsed as string "[]" instead of actual array [].

<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

This pull request addresses a bug in the GLM-4.7 MoE Detector where complex JSON Schema types were not being parsed correctly. Specifically, when a tool parameter was defined with complex structures like `anyOf` (e.g., `{"anyOf": [{"type": "array"}, {"type": "null"}]}`), the values were being treated as strings instead of their actual types (like arrays). For example, an array value `[]` would be parsed as the string `"[]"` rather than the actual array `[]`.

## Modifications

- Modified `get_argument_type` function in `glm47_moe_detector.py` to correctly parse complex JSON Schema structures including anyOf/oneOf/allOf and OpenAI-style array types like ["string", "null"]
- Improved `_format_value_complete` function to properly handle object/array types as JSON rather than strings
- Added additional parsing strategies in `parse_arguments` for object/array type handling
- Added comprehensive test cases in `test_function_call_parser.py` to verify complex type definitions are correctly parsed as their appropriate data types (e.g., arrays) rather than strings

## Accuracy Tests

The fix has been validated with test cases that confirm complex JSON Schema types are now correctly parsed. The tests verify that array values like `[]` are correctly interpreted as arrays instead of strings.

## Benchmarking and Profiling

No performance impact is expected from these changes as they only affect the parsing logic when complex types are encountered.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

JustinTong0323 commented:
/tag-and-rerun-ci

wwzeng1 commented:
Hi, I tried this and this PR introduces a bug where a partial tool call can be early matched and incorrectly returned. For example if the tool should be `<tool_call>read_file` you'll early match `<tool_call>read` which causes the tool to fail.

I'll leave a separate comment with the fix

Leoyzen commented on python/sglang/srt/function_call/glm47_moe_detector.py:
```diff
@@ -408,17 +487,33 @@ def parse_streaming_increment(
+                func_args_raw = (
+                    func_args_raw.strip() if func_args_raw is not None else ""
+                )
                 is_tool_end = partial_match.group(3)
```
@wwzeng1 The bug was introduced by original implementation #15333:
```python
        try:
            # Try to match a partial or complete tool call
            partial_match = re.search(
                pattern=r"<tool_call>(.*?)(<arg_key>.*?)?(</tool_call>|$)",
                string=current_text,
                flags=re.DOTALL,
            )
            if partial_match:
                func_name = partial_match.group(1).strip()
                func_args_raw = partial_match.group(2).strip()
                is_tool_end = partial_match.group(3)

                # Initialize state if this is the first tool call
                if self.current_tool_id == -1:
                    self.current_tool_id = 0
                    self.prev_tool_call_arr = []
                    self.streamed_args_for_tool = [""]
                    self._streamed_raw_length = 0
                    self.current_tool_name_sent = False
                    self._reset_streaming_state()

                # Ensure we have enough entries in our tracking arrays
                while len(self.prev_tool_call_arr) <= self.current_tool_id:
                    self.prev_tool_call_arr.append({})
                while len(self.streamed_args_for_tool) <= self.current_tool_id:
                    self.streamed_args_for_tool.append("")

                # Send tool name first if not sent yet
                if not self.current_tool_name_sent:
                    assert func_name, "func_name should not be empty"
                    calls.append(
                        ToolCallItem(
                            tool_index=self.current_tool_id,
                            name=func_name,
                            parameters="",
                        )
                    )
                    self.current_tool_name_sent = True
                    self._streamed_raw_length = 0
                    self._reset_streaming_state()
                    # Store the tool call info
                    self.prev_tool_call_arr[self.current_tool_id] = {
                        "name": func_name,
                        "arguments": {},
                    }
```

When the model output partial function name tokens, the regex still match, and send the function name, then set `current_tool_name_sent=True`.

This will cause the remain function name will not being send after.

Leoyzen committed:
fix: implement complex JSON Schema type inference for GLM function call detectors

This commit implements RFC: improve type inference for complex JSON Schema
(see plans/rfc_improve_type_inference_for_complex_json_schema.md)

## Problem

Previously, GLM function call detectors (glm4_moe_detector.py and
glm47_moe_detector.py) only read direct "type" field from JSON
Schema, unable to handle complex type constraints:

- anyOf: parameter can be any of multiple types
- oneOf: parameter must be one of multiple types
- allOf: parameter must satisfy all type definitions
- enum: parameter must be one of enum values

This caused type inference to return None for these complex schemas,
forcing parser to rely on unreliable heuristic detection.

## Solution

1. Added `infer_type_from_json_schema()` function in utils.py with
   multi-level type inference following RFC specification:
   - Priority 1: Direct type field (including type arrays)
   - Priority 2: anyOf/oneOf - analyze all options, return safest type
   - Priority 3: enum - infer type from enum values
   - Priority 4: allOf - merge all type constraints
   - Priority 5: properties - infer as object
   - Priority 6: items - infer as array

2. Updated `get_argument_type()` in both glm4_moe_detector.py and
   glm47_moe_detector.py to use new type inference function

3. Updated test expectations in test_glm47_moe_detector.py to reflect
   correct behavior (tests previously expected None to demonstrate issue)

## Changes

- python/sglang/srt/function_call/utils.py: New function
- python/sglang/srt/function_call/glm4_moe_detector.py: Updated
- python/sglang/srt/function_call/glm47_moe_detector.py: Updated
- test/registered/function_call/test_glm47_moe_detector.py: Updated tests

## Impact

- Fixes complex JSON Schema support in GLM detectors
- All 45 tests now pass (6 previously failed)
- Maintains backward compatibility with simple schemas
- Improves streaming parsing accuracy by eliminating type inference errors

Co-authored-by: SGLang Team

Leoyzen committed:
refactor: translate Chinese comments to English in infer_type_from_json_schema

Updated all Chinese comments and docstrings in the
infer_type_from_json_schema function to English for better
code maintainability and internationalization.

Changes:
- python/sglang/srt/function_call/utils.py: Updated comments

Co-authored-by: SGLang Team

CatherineSue commented:
LGTM. Could you take a look at CI failures?

Leoyzen commented:
@JustinTong0323  I'm not sure how to fix these failures. I checked the details, but the errors seem unrelated to the changes in this PR. What should I do?

Leoyzen committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix-complex-json

Leoyzen committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix-complex-json

Leoyzen commented:
@JustinTong0323 @CatherineSue Could you take a look about this PR?
